docs/developer-docs/docs/build-a-keychain/sdk/tutorial.md

```diff
@@ -0,0 +1,409 @@
+   - This allows for secure transactions and operations within the Warden Protocol.
+
+3. Integration with Blockchain:
+   - The service is configured to interact with a specific blockchain (identified by the ChainID).
+   - It uses gRPC to communicate with the blockchain node.
+
+4. Batching and Gas Management:
+   - The service batches operations and manages gas usage to optimize transaction costs and efficiency.
+
+In a production environment, you would need to implement actual key generation and signing logic, integrate with a secure key storage solution, and add more robust error handling and security measures.
+
+## Prerequisites
+
+`Go` (version 1.23 or later) installed on your system
+
+## Setting Up the Project
+
+1. Create a new directory for your project:
+
+    ```bash
+    mkdir warden-keychain-service && cd warden-keychain-service
+    ```
+
+2. Initialize a new Go module:
+
+   ```go
+   go mod init warden-keychain-service
+   ```
+
+3. Install the required dependencies:
+
+   ```go
+   go get github.com/warden-protocol/wardenprotocol/keychain-sdk
+   go get github.com/stretchr/testify
+   ```
+
+## Creating the Main Application
+
+1. Create a new file named `main.go` in your project directory and open it in your preferred text editor.
+
+2. Add the following code to `main.go` with the skeleton:
+
+```go
+package main
+
+import (
+"context"
+"log/slog"
+"os"
+"time"
+
+"github.com/warden-protocol/wardenprotocol/keychain-sdk"
+)
+
+func main() {
+// Set up a logger for debugging
+
+// Create a new keychain application
+
+})
+
+// Set up handlers for key requests and sign requests
+app.SetKeyRequestHandler(handleKeyRequest)
+app.SetSignRequestHandler(handleSignRequest)
+
+
+// Start the application
+
+
+// handleKeyRequest processes incoming key requests
+func handleKeyRequest(w keychain.KeyResponseWriter, req *keychain.KeyRequest) {
+}
+
+// handleSignRequest processes incoming sign requests
+func handleSignRequest(w keychain.SignResponseWriter, req *keychain.SignRequest) {
+}
+```
+
+Let us first define `handleKeyRequest` function. This function takes in a `KeyResponseWriter` and a `KeyRequest` as parameters.Inside the function, let us create a logger using `slog.Default()` and log informational messages with the request ID and key type.
+
+**Note:** This function will create a dummy public key as a byte slice.
+
+Finally, let us call the `Fulfil` method on the `KeyResponseWriter` with the dummy public key. If there is an error, it logs an error message and calls the Reject method on the `KeyResponseWriter` with an error message.
+
+```go
+// handleKeyRequest processes incoming key requests
+func handleKeyRequest(w keychain.KeyResponseWriter, req *keychain.KeyRequest) {
+logger := slog.Default()
+logger.Info("received key request", "id", req.Id, "key_type", req.KeyType)
+
+// In a real application, you would generate a public key here
+// For this example, we'll use a dummy public key
+publicKey := []byte("dummy_public_key")
```

Comment:
this tx will fail on-chain because there is a check for the length of the public key